File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.java
import java.util.*;
class Solution {
    public int solution(int[] players, int m, int k) {
        int answer = 0;
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingInt(a->a[0]));
        int cnt = 0;
        int now = 0;
        for(int i=0; i<24;i++){
            while(!pq.isEmpty()&&pq.peek()[0]==i){
                now -= pq.poll()[1];
            }
            int need = players[i]/ m;
            int diff = now - need;
            
            if(diff<0){
                diff = -diff;
                now += diff;
                cnt+=diff;
                pq.offer(new int[]{i+k, diff});
                
            }
        }
        
        return cnt;
    }
}

File: 프로그래머스/2/389479. 서버 증설 횟수/서버 증설 횟수.py
import heapq
def solution(players, m, k):
    answer = 0
    
    q = []
    #현재 서버
    now_server = 0
    for i in range(24):

        while q and q[0][0] == i:
            now_server -= heapq.heappop(q)[1]
           
        need_server = players[i]//m
        tmp = now_server - need_server
        
        if(tmp<0):
            tmp = -tmp
            now_server += tmp
            answer += tmp
            heapq.heappush(q, (i+k,tmp))
    return answer